File: java-springboot-js/src/test/java/com/payneteasy/hostedfields/ControlTest.java
package com.payneteasy.hostedfields;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.Test;

/**
 * The 3DS callback checksum, on the vectors every other example uses. They are written out rather
 * than recomputed, so a change of algorithm here cannot quietly agree with itself.
 */
class ControlTest {

    private static final String MERCHANT_CONTROL = "test-merchant-control";

    /** sha1("approved" + "1234567" + "hf-abc" + "test-merchant-control") */
    private static final String CONTROL = "652ace404c4dfe8bba069ecee594ec23a89340e1";

    private static Map<String, String> callback() {
        Map<String, String> form = new HashMap<>();
        form.put("status", "approved");
        form.put("orderid", "1234567");
        form.put("merchant_order", "hf-abc");
        form.put("control", CONTROL);
        return form;
    }

    @Test
    void signsTheDocumentedFields() {
        assertEquals(CONTROL, Control.checksum("approved", "1234567", "hf-abc", MERCHANT_CONTROL));
    }

    @Test
    void acceptsASignedCallback() {
        assertTrue(Control.valid(callback(), MERCHANT_CONTROL));
    }

    @Test
    void rejectsAnEditedField() {
        for (String name : new String[] {"status", "orderid", "merchant_order"}) {
            Map<String, String> form = callback();
            form.put(name, form.get(name) + "x");
            assertFalse(Control.valid(form, MERCHANT_CONTROL), "an edited " + name + " must not verify");
        }
    }

    @Test
    void rejectsAWrongControlOfTheRightLength() {
        Map<String, String> form = callback();
        form.put("control", "0".repeat(CONTROL.length()));
        assertFalse(Control.valid(form, MERCHANT_CONTROL));
    }

    /** A forged control of the wrong length has to be a 403, not an exception on the way to a 500. */
    @Test
    void rejectsAControlOfTheWrongLength() {
        for (String forged : new String[] {"", "short", CONTROL + "extra"}) {
            Map<String, String> form = callback();
            form.put("control", forged);
            assertFalse(Control.valid(form, MERCHANT_CONTROL));
        }
    }

    /** A callback with nothing in it still has a well-defined checksum, and no control at all
     *  never verifies. */
    @Test
    void emptyCallbackStillHasAChecksum() {
        assertFalse(Control.valid(Map.of(), MERCHANT_CONTROL));

        // sha1("" + "" + "" + "test-merchant-control")
        assertEquals("1a66987ac24e927ff2979f83a41cb818936a9e62", Control.checksum("", "", "", MERCHANT_CONTROL));
        assertTrue(Control.valid(Map.of("control", "1a66987ac24e927ff2979f83a41cb818936a9e62"), MERCHANT_CONTROL));
    }
}

File: kotlin-ktor-js/src/test/kotlin/com/payneteasy/hostedfields/ControlTest.kt
package com.payneteasy.hostedfields

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * The 3DS callback checksum, on the same vectors as every other example.
 *
 * No credentials and no stubbed environment: [Control] is pure and [Settings] is built in `main`,
 * so there is nothing to stub.
 */
class ControlTest {
    private val merchantControl = "test-merchant-control"

    // sha1("approved" + "1234567" + "hf-abc" + "test-merchant-control")
    private val control = "652ace404c4dfe8bba069ecee594ec23a89340e1"

    private fun signedCallback() =
        mutableMapOf(
            "status" to "approved",
            "orderid" to "1234567",
            "merchant_order" to "hf-abc",
            "control" to control,
        )

    @Test
    fun `signs the documented fields`() {
        assertEquals(control, Control.checksum("approved", "1234567", "hf-abc", merchantControl))
    }

    @Test
    fun `accepts a signed callback`() {
        assertTrue(Control.valid(signedCallback(), merchantControl))
    }

    @Test
    fun `rejects an edited field`() {
        for (field in listOf("status", "orderid", "merchant_order")) {
            val tampered = signedCallback()
            tampered[field] = tampered.getValue(field) + "-edited"
            assertFalse(Control.valid(tampered, merchantControl), "$field is not covered by the checksum")
        }
    }

    @Test
    fun `rejects a wrong control of the right length`() {
        val tampered = signedCallback()
        tampered["control"] = "0".repeat(control.length)
        assertFalse(Control.valid(tampered, merchantControl))
    }

    @Test
    fun `rejects a control of the wrong length without throwing`() {
        // A hand-edited return URL has to be a 403, not an exception on the way to a 500
        for (wrong in listOf("", "short", control + "extra")) {
            val tampered = signedCallback()
            tampered["control"] = wrong
            assertFalse(Control.valid(tampered, merchantControl))
        }
    }

    @Test
    fun `an empty callback still has a checksum`() {
        // Which is what stops a bare /result with no control reaching the page
        assertFalse(Control.valid(emptyMap(), merchantControl))
        val empty = mapOf("control" to Control.checksum("", "", "", merchantControl))
        assertTrue(Control.valid(empty, merchantControl))
    }
}
